File: examples/C/udpping2.c
//  UDP ping command
//  Model 2, uses separate UDP library

#include <czmq.h>
#include "udplib.c"
#define PING_PORT_NUMBER 9999
#define PING_MSG_SIZE    1
#define PING_INTERVAL    1000  //  Once per second

int main (void)
{
    zctx_t *ctx = zctx_new ();
    udp_t *udp = udp_new (PING_PORT_NUMBER);
    
    byte buffer [PING_MSG_SIZE];
    zmq_pollitem_t pollitems [] = {
        { NULL, udp_handle (udp), ZMQ_POLLIN, 0 }
    };
    //  Send first ping right away
    uint64_t ping_at = zclock_time ();
    
    while (!zctx_interrupted) {
        long timeout = (long) (ping_at - zclock_time ());
        if (timeout < 0)
            timeout = 0;
        if (zmq_poll (pollitems, 1, timeout * ZMQ_POLL_MSEC) == -1)
            break;              //  Interrupted

        //  Someone answered our ping
        if (pollitems [0].revents & ZMQ_POLLIN)
            udp_recv (udp, buffer, PING_MSG_SIZE);
        
        if (zclock_time () >= ping_at) {
            puts ("Pinging peers...");
            buffer [0] = '!';
            udp_send (udp, buffer, PING_MSG_SIZE);
            ping_at = zclock_time () + PING_INTERVAL;
        }
    }
    udp_destroy (&udp);
    zctx_destroy (&ctx);
    return 0;
}

File: examples/Java/Udpping2.java
import org.zeromq.ZContext;
import org.zeromq.ZMQ;

import java.io.IOException;
import java.net.SocketAddress;
import java.nio.ByteBuffer;

/**
 * UDP ping command
 * Model 2, uses separate UDP library
 */
public class Udpping2 {
  private static final int PING_PORT_NUMBER = 9999;
  private static final int PING_MSG_SIZE = 1;
  private static final int PING_INTERVAL = 1000; //  Once per second

  public static void main(String[] args) throws IOException {
    ZContext ctx = new ZContext();
    Udplib udp = new Udplib(PING_PORT_NUMBER);

    ByteBuffer buffer = ByteBuffer.allocate(PING_MSG_SIZE);
    ZMQ.PollItem[] pollItems = new ZMQ.PollItem[]{
            new ZMQ.PollItem(udp.handle(), ZMQ.Poller.POLLIN)};
    //  Send first ping right away
    long pingAt = System.currentTimeMillis();

    while (!Thread.currentThread().isInterrupted()) {
      long timeout = pingAt - System.currentTimeMillis();
      if (timeout < 0)
        timeout = 0;
      if (ZMQ.poll(pollItems, 1, timeout) == -1)
        break;              //  Interrupted

      //  Someone answered our ping
      if (pollItems[0].isReadable()){
        SocketAddress sourceAddr = udp.recv(buffer);
        System.out.printf("new peer found: %s\n",sourceAddr);
      }

      if (System.currentTimeMillis() >= pingAt) {
        System.out.println("Pinging peers…");
        buffer.put(0, (byte) '!');
        udp.send(buffer);
        pingAt = System.currentTimeMillis() + PING_INTERVAL;
      }
    }
    udp.destroy();
    ctx.destroy();
  }
}
